lib/dialects/abstract/query-generator/composition.js

```diff
@@ -0,0 +1,147 @@
+'use strict';
+
+const { CompositionError } = require('../../../errors');
+const Utils = require('../../../utils');
+
+/*
+ * Self contained value representation
+ *
+ * It can be converted to an escaped sql string or to a binded param. It also
+ * must be self contained, the escaper or binder should be part of the slot. It
+ * is identified for being an instace of this class, otherwise reject (reflects
+ * unsafe handling)
+ *
+ */
+class Slot {
+  constructor(value, field, options) {
+    this.value = value;
+    this.field = field;
+    this.options = options || {};
+  }
+}
+
+/*
+ * It is a placeholder in a QueryItems instance
+ *
+ */
+class Placeholder {
+  constructor(name) {
+    this.name = name;
+  }
+}
+
+/*
+  Buildind block for queries of any dialect
+
+  Should be simple, secure and composable. It can hold three items: (sql) strings,
+  value *slots* or placeholders. Keep this class simple, for higher abstractions
+  build something else. This should not depend on dialect or db version.
+
+  @private
+ */
+class Composition {
+  constructor(...items) {
+    // Should clone *items* array. Should not copy *items* array by reference.
+    this.items = this.constructor._processArgs(items);
+  }
+  get length() {
+    return this.items.length;
+  }
+  static _processArgs(items) {
```

Comment:
Note: This probably should be not static since you rely on inheritance here, eg. you use `this.constructor._processArgs` below.
There isn't much of a benefit to making it static.

Reply:
I usually make static all functions that don't expect an instance. ES6 classes inherit static methods, so inheritance should not be a problem.

Reply:
The point is that you are effectively using a static method like an instance method, it doesn't really matter that there is no `this` context (since `this` would just be the class itself, I'd just ask you to make it non static :)